src/async/ExtensiblePromise.ts

```diff
@@ -124,8 +126,8 @@ export default class ExtensiblePromise<T> {
 	 *
 	 * @returns {ExtensiblePromise}
 	 */
-	catch<U>(onRejected: (reason: Error) => (U | Thenable<U>)): this;
-	catch<U>(onRejected: (reason: Error) => void): this {
+	catch<U>(onRejected: (reason: Error) => (U | Thenable<U>)): ExtensiblePromise<U>;
+	catch<U>(onRejected: (reason: Error) => void): ExtensiblePromise<U> {
```

Comment:
I think there is a problem with the types here still. Consider the following example:

```typescript
import ExtensiblePromise from 'dojo-core/async/ExtensiblePromise';

export function booleanPromise(): ExtensiblePromise<boolean> {
	const promise = new ExtensiblePromise(resolve => {
		resolve(true);
	});

	return promise.catch(error => { console.log(error); });
}
```

In this example, the `catch` call is changing the promise type from `ExtensiblePromise<boolean>` to `ExtensiblePromise<void>`, which should not be correct.

I think the following method type definition should work:

```typescript
	catch(onRejected: (reason: Error) => T | Thenable<T> | void): ExtensiblePromise<T>;
	catch<U>(onRejected: (reason: Error) => U | Thenable<U>): ExtensiblePromise<U> {
```